server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -308,21 +355,42 @@ public void putSynonymsSet(String synonymSetId, SynonymRule[] synonymsSet, Actio
     }
 
     public void putSynonymRule(String synonymsSetId, SynonymRule synonymRule, ActionListener<SynonymsReloadResult> listener) {
-        checkSynonymSetExists(synonymsSetId, listener.delegateFailure((l1, obj) -> {
-            try {
-                IndexRequest indexRequest = createSynonymRuleIndexRequest(synonymsSetId, synonymRule).setRefreshPolicy(
-                    WriteRequest.RefreshPolicy.IMMEDIATE
-                );
-                client.index(indexRequest, l1.delegateFailure((l2, indexResponse) -> {
-                    UpdateSynonymsResultStatus updateStatus = indexResponse.status() == RestStatus.CREATED
-                        ? UpdateSynonymsResultStatus.CREATED
-                        : UpdateSynonymsResultStatus.UPDATED;
-
-                    reloadAnalyzers(synonymsSetId, false, l2, updateStatus);
+        checkSynonymSetExists(synonymsSetId, listener.delegateFailureAndWrap((l1, obj) -> {
+            // Count synonym rules to check if we're at maximum
+            client.prepareSearch(SYNONYMS_ALIAS_NAME)
+                .setQuery(
+                    QueryBuilders.boolQuery()
+                        .must(QueryBuilders.termQuery(SYNONYMS_SET_FIELD, synonymsSetId))
+                        .filter(QueryBuilders.termQuery(OBJECT_TYPE_FIELD, SYNONYM_RULE_OBJECT_TYPE))
+                )
+                .setSize(0)
+                .setPreference(Preference.LOCAL.type())
+                .setTrackTotalHits(true)
+                .execute(l1.delegateFailureAndWrap((searchListener, searchResponse) -> {
+                    long synonymsSetSize = searchResponse.getHits().getTotalHits().value;
+                    if (synonymsSetSize >= MAX_SYNONYMS_SETS) {
+                        // We could potentially update a synonym rule when we're at max capacity, but we're keeping this simple
```

Comment:
We could consider supporting updates here by adding a `must_not` clause to the query on the synonym ID? Then it would be below the max if it existed.

Reply:
That's a good way of doing this! Again, I feel it a bit complicated for such an edge case.

Reply:
Agreed, it's not really an edge case to be at capacity and to update the existing rules. I think we need a more robust approach here like executing all updates on the master node and executing the updates sequentially.

Reply:
> it's not really an edge case to be at capacity and to update the existing rules

This edge case happens only if we're updating an individual synonym rule when we're at max capacity. Updating rules in batch means that first we remove all the rules and apply the updates in a bulk request.

It sounds weird to me to update individual rules when we have 100k synonyms in a synonym set- at that scale, I would think that users do batch updates of rules, which effectively replace the existing ones.

But if y'all think this needs to be dealt with, I will! I'll update the code and ping when ready 👍 

> I think we need a more robust approach here like executing all updates on the master node and executing the updates sequentially

So do you think that this should be a `TransportMasterNodeAction`? As this updates an index and not the cluster state, what would be the advantages of applying the action on the master node? 

As we're doing a bulk request under the hood, doesn't that mean that we're applying the updates sequentially?

Reply:
> Updating rules in batch means that first we remove all the rules and apply the updates in a bulk request.

This is another case of non-consistent updates, we need to ensure some ordering here.

> So do you think that this should be a TransportMasterNodeAction? As this updates an index and not the cluster state, what would be the advantages of applying the action on the master node?

Having a single place where updates can occur but as you noticed that won't be enough. We also need to ensure that all updates are done sequentially, not in a single bulk request but globally when multiple synonyms updates are done in parallel. See the `MetadataMappingService` for an example of service that applies updates sequentially.

Reply:
Thanks @jimczi , I will take a look into that.

I think we should create a separate issue for ensuring updates are applied sequentially, as this is not directly related to this change?

Reply:
I think it's related, we cannot limit the number of synonyms without ensuring that updates are applied sequentially.